libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,589 @@
+#pragma once
+
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/webassembly/common.hpp>
+#include <eosio/chain/webassembly/eos-vm.hpp>
+#include <eosio/chain/webassembly/preconditions.hpp>
+#include <fc/crypto/sha1.hpp>
+
+namespace eosio { namespace chain { namespace webassembly {
+
+   template <auto HostFunction, typename... Preconditions>
+   struct host_function_registrator {
+      constexpr host_function_registrator(const std::string& mod_name, const std::string& fn_name) {
+         using rhf_t = eos_vm_host_functions_t;
+         rhf_t::add<HostFunction, Preconditions...>(mod_name, fn_name);
+      }
+   };
+
+#define REGISTER_INJECTED_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, ##__VA_ARGS__> NAME ## _registrator = {eosio_injected_module_name, #NAME};
+
+#define REGISTER_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, static_check_wl_args, context_aware_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_CF_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, static_check_wl_args, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, context_aware_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_CF_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_CF_ONLY_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, context_free_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+   class interface {
+      public:
+         interface(apply_context& ctx) : context(ctx) {}
+
+         inline apply_context& get_context() { return context; }
+         inline const apply_context& get_context() const { return context; }
+
+         // checktime api
```

Comment:
with wabt gone for good, seems like checktime() and call_depth_assert() can be removed?